File: problems/46-Permutations/solution.go
func traverse(nums []int, current *[]int, result *[][]int) {
	if len(nums) == 1 {
		*current = append(*current, nums[0])
		newResult := make([]int, len(*current))
		copy(newResult, *current)
		*result = append(*result, newResult)
		*current = (*current)[:len(*current)-1]
		return
	}
	for i := 0; i < len(nums); i++ {
		*current = append(*current, nums[i])

		newNums := make([]int, i)
		copy(newNums, nums[:i])
		newNums = append(newNums, nums[i+1:]...)

		traverse(newNums, current, result)
		*current = (*current)[:len(*current)-1]
	}
}

func permute(nums []int) [][]int {
	result := make([][]int, 0)
	current := make([]int, 0, len(nums))

	traverse(nums, &current, &result)

	return result
}

File: problems/46-Permutations/solution.java
class Solution {
    private void backtrace(LinkedList<Integer> curr, LinkedList<Integer> remain, List<List<Integer>> result) {
        if (remain.size() == 1) {
            curr.add(remain.getFirst());
            result.add(new LinkedList(curr));
            curr.removeLast();

            return;
        }

        for (int i = 0; i < remain.size(); i++) {
            Integer removed = remain.remove(i);
            curr.add(removed);
            backtrace(curr, remain, result);
            curr.removeLast();
            remain.add(i, removed);
        }
    }
    
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new LinkedList<>();

        LinkedList<Integer> curr = new LinkedList<>(), remain = new LinkedList<>();
        for (int num : nums) {
            remain.add(num);
        }
        
        backtrace(curr, remain, result);

        return result;
    }
}

